# PhD_scripts
Scripts used for data analysis in the MHC introgression project

1. AmpliSAS_running/ - scripts used to run AmpliSAS software
2. AmpiSAS_resultAnalysis/ - scripts used to analyse AmpliSAS outputs
3. BLASTing/ - scripts used to run and analyse BLAST searches
4. short_simple_specific/ - simple scripts created to perform short and specific tasks
5. hzar_wrapper/ - scripts used to fit and analyse geographic clines with hzar package
6. genomicClines/ - script used to fit genomic clines with ML method and gather their 2LL intervals
7. Direct_analysis_MHCintrogression/ - scripts and data used to analyse MHC introgression
8. build_circular_tree.R - Draws circular tree of DNA sequences - filters sequences, aligns sequences, builds tree, draws circular tree
9. MHC_allele_sharing_with_permutations_TG_SES.py - modified prof. Babik's script performing permutation test for increased sharing of MHC alleles in parapatry. 
10. primers_design_MHC.py - script to design multiple primers for highly variable regions (e.g. MHC)
11. raw_mapping.py - perform mapping of raw reads to reference sequences using bowtie2
12. sra.tool.py - download raw reads from SRA based on the text file with accessions
13. perform STRUCTURE analysis in parallel
14. trinity_tool.py - creates de novo RNA assembly using Trinity software

